src/Microsoft.Extensions.Diagnostics.HealthChecks/HealthCheckService.cs

```diff
@@ -0,0 +1,118 @@
+
+        /// <summary>
+        /// Constructs a <see cref="HealthCheckService"/> from the provided collection of <see cref="IHealthCheck"/> instances, and the provided logger.
+        /// </summary>
+        /// <param name="healthChecks">The <see cref="IHealthCheck"/> instances that have been registered in the application.</param>
+        /// <param name="logger">A <see cref="ILogger{T}"/> that can be used to log events that occur during health check operations.</param>
+        public HealthCheckService(IEnumerable<IHealthCheck> healthChecks, ILogger<HealthCheckService> logger)
+        {
+            // Scan the list for duplicate names to provide a better error if there are duplicates.
+            var names = new HashSet<string>();
+            var duplicates = new List<string>();
+            foreach (var check in healthChecks)
+            {
+                if(!names.Add(check.Name))
+                {
+                    duplicates.Add(check.Name);
+                }
+            }
+
+            if(duplicates.Count > 0)
+            {
+                throw new InvalidOperationException($"Duplicate health checks were registered with the name(s): {string.Join(", ", duplicates)}");
+            }
+
+            Checks = healthChecks.ToDictionary(c => c.Name);
+            _logger = logger;
+
+            if (_logger.IsEnabled(LogLevel.Debug))
+            {
+                foreach (var check in Checks)
+                {
+                    _logger.LogDebug("Health check '{healthCheckName}' has been registered", check.Key);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Runs all the health checks in the application and returns the aggregated status.
+        /// </summary>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> which can be used to cancel the health checks</param>
+        /// <returns>
+        /// A <see cref="Task{T}"/> which will complete when all the health checks have been run,
+        /// yielding a <see cref="CompositeHealthCheckResult"/> containing the results.
+        /// </returns>
+        public async Task<CompositeHealthCheckResult> CheckHealthAsync(CancellationToken cancellationToken = default)
+        {
+            var results = new Dictionary<string, HealthCheckResult>(Checks.Count);
+            foreach (var pair in Checks)
+            {
+                // If the health check does things like make Database queries using EF or backend HTTP calls,
+                // it may be valuable to know that logs it generates are part of a health check. So we start a scope.
+                using (_logger.BeginScope(new HealthCheckLogScope(pair.Key)))
+                {
+                    HealthCheckResult result;
+                    try
+                    {
+                        _logger.LogTrace("Running health check: {healthCheckName}", pair.Key);
+                        result = await pair.Value.CheckStatusAsync(cancellationToken);
```

Comment:
Should there be a configurable timeout per check? Though I guess you could do that on the options for that check.

Reply:
Also more complicated than we want to do right now. That's what the sample had, and it got messy and started interfering with everything. Checks don't cache ever in this model, only the middleware caches the check. It could have per-check customization for caching, but that's rarely something you want to do. None of the other systems that I've looked at really do any caching beyond a simple TTL for the full results.

Reply:
If a health check wants to do timeout, they can add that functionality. Adding it to the outer system will add fairly significant complication.